src/wake_on.rs

```diff
@@ -0,0 +1,117 @@
+//! Smart pointers to wake tasks on access
+use async_std::task::Waker;
+use std::ops::{Deref, DerefMut};
+
+/// A wrapper type which wakes tasks whenever the wrapped value is accessed
+/// through an `&mut` reference.
+///
+/// `T` is the type of the value being wrapped. This struct is `Deref` and
+/// `DerefMut` for that type, giving `&T` and `&mut T` respectively.
+/// When a `Waker` is registered with `set_waker`, that `Waker` is woken
+/// whenever the wrapped value is accessed through an `&mut` reference
+/// and therefore potentially mutated.
+///
+/// This is useful when there is a future polling the state of the wrapped
+/// value. It needs to be awoken whenever that value changes so that they
+/// can check whether or not its value is in a state that will let the
+/// future make progress. That future can register the `Waker` from the
+/// `Context` it is passed with the `WakeOnWrite` wrapping the value it is
+/// interested in so that all mutations cause it to be woken going forward.
+///
+/// This type isn't effective for observing changes on values with interior
+/// mutablity, because it only wakes on `&mut` access.
+#[derive(Default, Debug, Clone)]
+pub struct WakeOnWrite<T> {
+    inner: T,
+    waker: Option<Waker>,
+}
+
+impl<T> WakeOnWrite<T> {
+    /// Create a new `WakeOnWrite` with the given value.
+    pub fn new(value: T) -> Self {
+        Self {
+            inner: value,
+            waker: None,
+        }
+    }
+
+    /// Set the `Waker` to be awoken when this value is mutated.
+    ///
+    /// Returns the currently registered `Waker`, if there is one.
+    pub fn set_waker(wow: &mut Self, waker: Waker) -> Option<Waker> {
+        wow.waker.replace(waker)
+    }
+
+    /// Removes and returns the currently registered `Waker`, if there is one.
+    pub fn take_waker(wow: &mut Self) -> Option<Waker> {
+        wow.waker.take()
+    }
+
```

Comment:
In using the library in qaul.net I've found this to be a convenient function to expose

```suggestion

    /// Invoke the waker, if it exists
    pub fn wake_if_waker(wow: &mut Self) {
        if let Some(w) = wow.waker {
            w.wake();
        }
    }
```

Reply:
I prefer just naming it `wake()`, but otherwise, good call. thank you!